### ISSUE TITLE
Some Questions of Loops step

### ISSUE THREAD

**Issue by myitroad** [JanusGraph/janusgraph#1257]
Hi there,
We have some questions of `loops step` in our project develop using JanusGraph. 

## questions
We want find some potential paths between two vertices, such as from vertex with property private_key=61656 to another vertex with property private_key=24672. We use `until step` and `loops step` to limit the traversal deeps. But we found the `loops step` seems doesn't work.

To illustrate this, there is a simple graph to present our scenario(here we use property `private_key` to present vertex): 
![61656-graph](https://user-images.githubusercontent.com/17537048/45533555-aa4b5100-b82a-11e8-8e83-44b319d937ee.png)

What we want is to find path between the red circled vertex and the green circled vertex, but the traverser seems goes to the yellow circled vertex!

We use traversal like this:
```
g.V().has("private_key", "61656").repeat(__.both().simplePath()).until(__.has("private_key", "24672").and().loops().is(P.lte(1))).path().toList();
```

We have traced the source code of JanusGraph, found loops increase in this method:
`org.apache.tinkerpop.gremlin.process.traversal.traverser.B_O_S_SE_SL_Traverser#incrLoops(final String stepLabel)`

Obviously we set loops to lte 1, but it doesn't go into effect.

We suppose that the traverser only have 3 paths to go, `61656->24672, 61656->4192, 61656->4320`. But in fact, we found the loops param and traverser times as list (here we use property `private_key` to present vertex):

| loops | times | path |
| ----- | ----- | ------ |
| 1 | 3 | 61656->24672, 61656->4192, 61656->4320 |
| 2 | 3 | 61656->4192->4296, 61656->4320->4296, 61656->4320->12368 |
| 3 | 8 | 61656->4192->4296->4320, 61656->4192->4296->other1, 61656->4192->4296->other2, 61656->4192->4296->other3, 61656->4320->4296->4192, 61656->4320->4296->other1, 61656->4320->4296->other2, 61656->4320->4296->other3 |

## Operation detail 
Below is concise process of the operation in gremlin-console:
- connect graph database
```
gremlin> graph = JanusGraphFactory.open('conf/gremlin-server/janusgraph-loops-test.properties')
```

- Create graph management and traversal
```
mgmt = graph.openManagement()
g=graph.traversal()
```

- Create vertex and edge label
```
person = mgmt.makeVertexLabel('person').make()
knows = mgmt.makeEdgeLabel('knows').multiplicity(MULTI).make()
private_key = mgmt.makePropertyKey('private_key').dataType(String.class).cardinality(Cardinality.SINGLE).make()
mgmt.commit()
```

- Vertices
```
v_61656 = g.addV('person').property("private_key", "61656");
v_24672 = g.addV('person').property("private_key", "24672");
v_4192 = g.addV('person').property("private_key", "4192");
v_4320 = g.addV('person').property("private_key", "4320");
v_4296 = g.addV('person').property("private_key", "4296");
v_12368 = g.addV('person').property("private_key", "12368");
v_other1 = g.addV('person').property("private_key", "other1");
v_other2 = g.addV('person').property("private_key", "other2");
v_other3 = g.addV('person').property("private_key", "other3");
```

- Edges
```
g.V().has("private_key","61656").next().addEdge('knows',g.V().has("private_key","24672").next())
g.V().has("private_key","61656").next().addEdge('knows',g.V().has("private_key","4192").next())
g.V().has("private_key","61656").next().addEdge('knows',g.V().has("private_key","4320").next())
g.V().has("private_key","4192").next().addEdge('knows',g.V().has("private_key","4296").next())
g.V().has("private_key","4320").next().addEdge('knows',g.V().has("private_key","4296").next())
g.V().has("private_key","4320").next().addEdge('knows',g.V().has("private_key","12368").next())
g.V().has("private_key","4296").next().addEdge('knows',g.V().has("private_key","other1").next())
g.V().has("private_key","4296").next().addEdge('knows',g.V().has("private_key","other2").next())
g.V().has("private_key","4296").next().addEdge('knows',g.V().has("private_key","other3").next())
```

- Commit transaction
```
graph.tx().commit()
```

- My traversal
```
g.V().has("private_key", "61656").repeat(__.both().simplePath()).until(__.has("private_key", "24672").and().loops().is(P.lte(1))).path().toList();
```

## `janusgraph-loops-test.properties` file
```
gremlin.graph=org.janusgraph.core.JanusGraphFactory
storage.backend=hbase

storage.hostname=host1,host2,host3
storage.hbase.table=janusgraph_loops_test
storage.hbase.ext.zookeeper.znode.parent=/hbase-unsecure

index.janusgraph_loops_test.backend=elasticsearch
index.janusgraph_loops_test.index-name=janusgraph_loops_test

index.janusgraph_loops_test.hostname=host1,host2,host3
```

**Comment by pluradj** [comment]
Thanks for the clear reproduce steps. I got the expected result using JanusGraph 0.2.1.

```
gremlin> g.V().has("private_key", "61656").
......1>   repeat(__.both().simplePath()).
......2>   until(__.has("private_key", "24672").and().loops().is(P.lte(1))).
......3>   path().by("private_key").toList()
==>[61656,24672]
```

https://gist.github.com/pluradj/56c22abc784e7750cbf7dcac30522655

@myitroad — please use GitHub only to report specific bugs or feature requests.

For general questions or open-ended discussion, please use either the Gitter channel or the `janusgraph-users@` mailing list: https://github.com/janusgraph/janusgraph/#community . This particular question actually on loop behavior isn't specific to JanusGraph and perhaps would be better directed towards the `gremlin-users` mailing list: https://groups.google.com/forum/#!forum/gremlin-users

**Comment by myitroad** [comment]
Thanks for you reply. @pluradj 

We can also got the expected result using JanusGraph 0.2.0. But, what we want to know is not only the correct result of the traversal, but also the traversal mechanism. 

When we trace the traversal steps, we found there are unnecessary process just as I commented above.

For example, We suppose the  reasonable traversal is only traverser on `61656->24672, 61656->4192, 61656->4320` when we limlit the loops to `lte(1)`. But actually, the traverser seems break the limit and does lots of extra jobs just as I described in above chart. 

We guess if there are something to do to optimize the traversal steps?

Best regards!
